services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_states.py

```diff
@@ -0,0 +1,90 @@
+"""
+States from Docker Tasks and docker Containers are mapped to ServiceState.
+"""
+import logging
+from typing import Dict, Set, Tuple
+
+from ...models.schemas.dynamic_services import ServiceState
+
+logger = logging.getLogger(__name__)
+
+TASK_STATES_FAILED: Set[str] = {"failed", "rejected"}
+TASK_STATES_PENDING: Set[str] = {"pending"}
+TASK_STATES_PULLING: Set[str] = {"assigned", "accepted", "preparing"}
+TASK_STATES_STARTING: Set[str] = {"ready", "starting"}
+TASK_STATES_RUNNING: Set[str] = {"running"}
+TASK_STATES_COMPLETE: Set[str] = {"complete", "shutdown"}
+
+
+TASK_STATES_ALL: Set[str] = (
+    TASK_STATES_FAILED
+    | TASK_STATES_PENDING
+    | TASK_STATES_PULLING
+    | TASK_STATES_STARTING
+    | TASK_STATES_RUNNING
+    | TASK_STATES_COMPLETE
+)
+
+
+# mapping container states into 4 categories
+CONTAINER_STATES_FAILED: Set[str] = {"restarting", "dead", "paused"}
+CONTAINER_STATES_PENDING: Set[str] = {"pending"}  # fake state
+CONTAINER_STATES_PULLING: Set[str] = {"pulling"}  # fake state
+CONTAINER_STATES_STARTING: Set[str] = {"created"}
+CONTAINER_STATES_RUNNING: Set[str] = {"running"}
+CONTAINER_STATES_COMPLETE: Set[str] = {"removing", "exited"}
+
+
+_TASK_STATE_TO_SERVICE_STATE: Dict[str, ServiceState] = {
+    **dict.fromkeys(TASK_STATES_FAILED, ServiceState.FAILED),
+    **dict.fromkeys(TASK_STATES_PENDING, ServiceState.PENDING),
+    **dict.fromkeys(TASK_STATES_PULLING, ServiceState.PULLING),
+    **dict.fromkeys(TASK_STATES_STARTING, ServiceState.STARTING),
+    **dict.fromkeys(TASK_STATES_RUNNING, ServiceState.RUNNING),
+    **dict.fromkeys(TASK_STATES_COMPLETE, ServiceState.COMPLETE),
+}
+
+_CONTAINER_STATE_TO_SERVICE_STATE: Dict[str, ServiceState] = {
+    **dict.fromkeys(CONTAINER_STATES_FAILED, ServiceState.FAILED),
+    **dict.fromkeys(CONTAINER_STATES_PENDING, ServiceState.PENDING),
+    **dict.fromkeys(CONTAINER_STATES_PULLING, ServiceState.PULLING),
+    **dict.fromkeys(CONTAINER_STATES_STARTING, ServiceState.STARTING),
+    **dict.fromkeys(CONTAINER_STATES_RUNNING, ServiceState.RUNNING),
+    **dict.fromkeys(CONTAINER_STATES_COMPLETE, ServiceState.COMPLETE),
+}
+
+
+def extract_task_state(task_status: Dict[str, str]) -> Tuple[ServiceState, str]:
+    last_task_error_msg = task_status["Err"] if "Err" in task_status else ""
+
+    task_state = _TASK_STATE_TO_SERVICE_STATE.get(
+        task_status["State"], ServiceState.STARTING
```

Comment:
hmm starting as default? how is this used? won't you get into an ever starting service in case of issue? can you check what might happen?

Reply:
this got refactored, no longer using a default